services/dynamic/dynamic-regex.service.js

```diff
@@ -0,0 +1,99 @@
+import Joi from 'joi'
+import RE2 from 're2'
+import { BaseService, InvalidParameter, queryParams } from '../index.js'
+import { url } from '../validators.js'
+import { renderDynamicBadge } from '../dynamic-common.js'
+
+export default class DynamicRegex extends BaseService {
+  static category = 'dynamic'
+  static route = {
+    base: `badge/dynamic/regex`,
+    pattern: '',
+    queryParamSchema: Joi.object({
+      url,
+      search: Joi.string().required(),
+      replace: Joi.string().optional(),
+      flags: Joi.string().optional(),
+      noMatch: Joi.string().optional(),
+    }),
+  }
+  static openApi = {
+    '/badge/dynamic/regex': {
+      get: {
+        summary: 'Dynamic Regex (re2) Badge [Experimental]',
+        description:
+          '⚠️ Experimental: This badge is considered experimental and may change or be removed at any time.\n\nThis badge will extract text from a file using re2 (a subset of regex: https://github.com/google/re2).\nThe main use-case is to extract values from unsupported plain-text files.\nFor example: if a file contains a line like `version - 2.4` you can extract the value `2.4` by using a search regex of `version - (.*)` and `$1` as replacement.\n\nFull Syntax documentation here: https://github.com/google/re2/wiki/Syntax',
+        parameters: queryParams(
+          {
+            name: 'url',
+            description:
+              'The URL to a file to search. The full raw content will be used as the search string.',
+            required: true,
+            example:
+              'https://raw.githubusercontent.com/badges/shields/refs/heads/master/README.md',
+          },
+          {
+            name: 'search',
+            description:
+              'A re2 expression that will be used to extract data from the document. Only the first matched text will be returned.',
+            required: true,
+            example: 'Every month it serves (.*?) images',
+          },
+          {
+            name: 'replace',
+            description:
+              'A regex expression that will be used as the replacement of the search regex, like `$1` to specify the first matched group, etc. If empty (default), no replacement will be done and the full matched text will be shown.',
+            required: false,
+            example: '$1',
+          },
+          {
+            name: 'flags',
+            description:
+              'Flags to be used when creating the regex, like `i` for case insensitive, or `m` for multiline. None by default.',
+            required: false,
+            example: 'imsU',
+          },
+          {
+            name: 'noMatch',
+            description:
+              'String to be returned if the regex does not match the input. Empty by default.',
+            required: false,
+          },
+        ),
+      },
+    },
+  }
+  static defaultBadgeData = { label: 'match' }
+
+  async handle(namedParams, { url, search, replace, flags, noMatch = '' }) {
```

Comment:
At the start of this function, can we validate that `flags` contains only valid characters (https://github.com/google/re2/blob/2024-07-02/doc/syntax.txt#L65-L70 ) and throw an `InvalidParameter` if it doesn't.

Reply:
Also can we link to
https://github.com/google/re2/blob/2024-07-02/doc/syntax.txt#L65-L70
in the docs so people know what flags can be used. The ones re2 supports are not the same as every regex engine e.g: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Regular_expressions#regex_flags

Reply:
Sure. Usually in these cases I prefer to leave the validation to the library, but since this is well documented a check should be good, and probably allow to differentiate invalid flag errors and invalid regex errors.

As for the documentation, I originally added a link to the specific documentation line in each parameter, but it was too much text and in the end I just left it only on the badge description.

Since in this case I'm going to add a custom check, do you think it may be a good idea to add all the flags in the parameter description instead of the link?  Or do you still prefer the link?

Reply:
It would be useful if the upstream library did some better validation here.
If I pass an invalid regex, we can just catch the syntax error and re-throw it as an `InvalidParameter()` but it seems like you can just pass any string as the flags argument and it will not be rejected.

I don't really mind if we link out to docs or reproduce them here, as long as it is clear what values are valid/invalid. I suggested https://github.com/google/re2/blob/2024-07-02/doc/syntax.txt#L65-L70 as a possible link because it doesn't seem to be easy to link to just the "flags" section on https://github.com/google/re2/wiki/Syntax but if there's another option, that's fine.

Reply:
I decided to include all the flags in the documentation (I purposely omitted the '-' possibility, since it's practically useless, but still check for its validity).

As for the link, that file you mention is the original unformatted text file, which may confuse as the information is the same as the other one, so I prefer to avoid linking it.
But, if you prefer, I'll have no other objections and I'll make the change.

Reply:
What you've done here looks good - thanks